MapboxNavigation/NavigationMapView.swift

```diff
@@ -899,40 +1035,122 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
+                stops.append([segmentStartPercentTraveled.nextUp: congestionColor])
+
+                let segmentEndPercentTraveled = CGFloat(1.0)
+                routeGradientStops.line[segmentEndPercentTraveled.nextDown] = congestionColor
+                continue
+            }
+
+            /**
+             If this is not the first or last congestion segment, then
+             the starting and ending percent values traveled for this segment
+             will be a fractional amount more/less than the actual values.
+             */
+            let segmentStartPercentTraveled = CGFloat((distanceTraveled / route.distance))
+            routeGradientStops.line[segmentStartPercentTraveled.nextUp] = congestionColor
+
+            distanceTraveled = distanceTraveled + distance
+
+            let segmentEndPercentTraveled = CGFloat((distanceTraveled / route.distance))
+            routeGradientStops.line[segmentEndPercentTraveled.nextDown] = congestionColor
+        }
+
+        /**
+         The casing layer around the route gradient is one single color,
+         so there's no need to give it the same stop/color combinations as
+         the main route stops dictionary.
+         */
+        routeGradientStops.casing[0.0] = routeCasingColor
+        routeGradientStops.casing[1.0] = routeCasingColor
     }
-    
-    func routeCasingStyleLayer(identifier: String, source: MGLSource) -> MGLStyleLayer {
-        let lineCasing = MGLLineStyleLayer(identifier: identifier, source: source)
-        
-        // Take the default line width and make it wider for the casing
-        lineCasing.lineWidth = NSExpression(format: "mgl_interpolate:withCurveType:parameters:stops:($zoomLevel, 'linear', nil, %@)", MBRouteLineWidthByZoomLevel.multiplied(by: 1.5))
-        
-        lineCasing.lineColor = NSExpression(forConditional: NSPredicate(format: "isAlternateRoute == true"),
-                                            trueExpression: NSExpression(forConstantValue: routeAlternateCasingColor),
-                                            falseExpression: NSExpression(forConstantValue: routeCasingColor))
-        
-        lineCasing.lineCap = NSExpression(forConstantValue: "round")
-        lineCasing.lineJoin = NSExpression(forConstantValue: "round")
-        
-        lineCasing.lineOpacity = NSExpression(forConditional: NSPredicate(format: "isAlternateRoute == true"),
-                                            trueExpression: NSExpression(forConstantValue: 1),
-                                            falseExpression: NSExpression(forConditional: NSPredicate(format: "isCurrentLeg == true"), trueExpression: NSExpression(forConstantValue: 1), falseExpression: NSExpression(forConstantValue: 0.85)))
-        
-        return lineCasing
+
+    /**
+     Given a congestion level, return its associated color.
+     */
+    private func getCongestionColor(for congestionLevel: String) -> UIColor {
```

Comment:
This method has no side effects and is synchronous, so rename it to `congestionColor(for:)`.